src/Tools/BuildValidator/CompilationDiff.cs

```diff
@@ -42,41 +51,89 @@ internal sealed class CompilationDiff
             string ILFilePath,
             string CustomDataFilePath);
 
-        public bool? AreEqual { get; }
-        public string OriginalPath { get; }
-        public ImmutableArray<Diagnostic> Diagnostics { get; }
+        private readonly ImmutableArray<Diagnostic> _diagnostics;
+        private readonly byte[]? _originalPortableExecutableBytes;
+        private readonly byte[]? _rebuildPortableExecutableBytes;
+        private readonly Compilation? _rebuildCompilation;
+        private readonly ImmutableArray<MetadataReferenceInfo> _references;
+        private readonly LocalReferenceResolver? _localReferenceResolver;
+        private readonly string? _message;
+
+        public AssemblyInfo AssemblyInfo { get; }
+        public CompilationOutcome Outcome { get; }
+
+        public bool Succeeded => Outcome == CompilationOutcome.Success;
 
-        private CompilationDiff(string originalPath, bool? areEqual)
+        public ImmutableArray<Diagnostic> Diagnostics
         {
-            AreEqual = areEqual;
-            OriginalPath = originalPath;
+            get
+            {
+                Debug.Assert(Outcome == CompilationOutcome.CompilationError);
```

Comment:
>Debug.Assert(Outcome == CompilationOutcome.CompilationError); [](start = 16, length = 61)

Why require the rebuild to fail?

Reply:
This is essentially a discriminated union hence this property is only valid in this state.

Reply:
> This is essentially a discriminated union hence this property is only valid in this state. 

It seems like the property returns a valid result in all cases, so it seems unnecessary to prevent calling this. And there may be valid reasons for calling this property regardless of the state, when expanding instances in the debugger for instance.

---
In reply to: [594672786](https://github.com/dotnet/roslyn/pull/51852#discussion_r594672786) [](ancestors = 594672786)